doc/manual/src/contributing/experimental-flags.md

```diff
@@ -0,0 +1,89 @@
+This section aims at describing the notion of a Nix "experimental
+flag", and how it fits into the big picture of the developpement of Nix.
+
+# What are these xp flags?
+
+Since Nix 2.4, some features of Nix can be flagged as
+experimental.Conceptually, this means that these features are not
+considered as part of Nix proper, but are only included in the code to
+make their developpement and testing easier.Practically, these require
+an explicit opt-in, by adding the name of the experimental feature to
+the `experimental-features` configuration option. This is really handy
+as it makes it harder to accidentally use an experimental feature
+without knowing it.
+
+# When should an experimental flag be used
+
+A change in the Nix codebase should be guarded by an experimental flag
+if it is likely to either introduce a regression of any kind, or if it
+changes the external interface of Nix (be it the language, the cli or
+anything else that's considered as "public").
```

Comment:
I don't think everything that changes the external interface requires an experimental feature. The main criterion is whether we're uncertain about whether the feature is a good idea at all, or about the exact interface, or the implementation is not sufficiently stable yet. But we wouldn't make (for instance) a new builtin function experimental unless we have doubts about it.

Reply:
Fair point. What about:

```suggestion
A change in the Nix codebase should be guarded by an experimental flag
if it is likely to either introduce a regression of any kind, or if it
implies a non-trivial change in the external interface of Nix (be it the language, the cli or
anything else that's considered as "public").
```
?

(For the specific “builtin” case, I’d actually be in favor of making them experimental because that doesn’t cost much and can avoid having to commit to them forever if they prove not to be that useful. But that’s not central)